app/notification.go

```diff
@@ -22,211 +22,247 @@ const (
-		updateMentionChans = append(updateMentionChans, a.Srv.Store.Channel().IncrementMentionCount(post.ChannelId, id))
+	if len(mentionedUsersList) != 0 {
+		message.Add("mentions", model.ArrayToJson(mentionedUsersList))
 	}
+	return message
+}
 
-	notification := &postNotification{
-		post:       post,
-		channel:    channel,
-		profileMap: profileMap,
-		sender:     sender,
+// Send or Register a notification as not sent based on user preference
+func (a *App) sendOrRegisterNotification(id string, channelMemberNotifyPropsMap map[string]model.StringMap, wasMentioned bool, status *model.Status,
+	post *model.Post, notification *postNotification, user *model.User, explicitMentions bool, channelWideMentions bool, replyToThreadType string) {
+
+	if ShouldSendPushNotification(user, channelMemberNotifyPropsMap[id], wasMentioned, status, post) {
+		a.sendPushNotification(
+			notification,
+			user,
+			explicitMentions,
+			channelWideMentions,
+			replyToThreadType,
+		)
+	} else {
+		// register that a notification was not sent
+		notificationRegistry := model.NotificationRegistry{
+			UserId:     id,
+			PostId:     post.Id,
+			SendStatus: model.PUSH_NOT_SENT,
+			Type:       model.PUSH_TYPE_MESSAGE,
+		}
+		_, appErr := a.Srv.Store.NotificationRegistry().Save(&notificationRegistry)
+		if appErr != nil {
+			mlog.Debug(appErr.Error())
+		}
 	}
+}
 
-	if *a.Config().EmailSettings.SendEmailNotifications {
-		for _, id := range mentionedUsersList {
-			if profileMap[id] == nil {
-				continue
-			}
-
-			userAllowsEmails := profileMap[id].NotifyProps[model.EMAIL_NOTIFY_PROP] != "false"
-			if channelEmail, ok := channelMemberNotifyPropsMap[id][model.EMAIL_NOTIFY_PROP]; ok {
-				if channelEmail != model.CHANNEL_NOTIFY_DEFAULT {
-					userAllowsEmails = channelEmail != "false"
-				}
-			}
-
-			// Remove the user as recipient when the user has muted the channel.
-			if channelMuted, ok := channelMemberNotifyPropsMap[id][model.MARK_UNREAD_NOTIFY_PROP]; ok {
-				if channelMuted == model.CHANNEL_MARK_UNREAD_MENTION {
-					mlog.Debug(fmt.Sprintf("Channel muted for user_id %v, channel_mute %v", id, channelMuted))
-					userAllowsEmails = false
-				}
-			}
-
-			//If email verification is required and user email is not verified don't send email.
-			if *a.Config().EmailSettings.RequireEmailVerification && !profileMap[id].EmailVerified {
-				mlog.Error(fmt.Sprintf("Skipped sending notification email to %v, address not verified. [details: user_id=%v]", profileMap[id].Email, id))
-				continue
-			}
-
-			var status *model.Status
-			var err *model.AppError
-			if status, err = a.GetStatus(id); err != nil {
-				status = &model.Status{
-					UserId:         id,
-					Status:         model.STATUS_OFFLINE,
-					Manual:         false,
-					LastActivityAt: 0,
-					ActiveChannel:  "",
-				}
-			}
+func (a *App) setStatus(id string, modelStatus string, manual bool, lastActivity int64, activeChannel string) *model.Status {
```

Comment:
Actually, this function doesn't need so much parameters, because always receive the same number of parameters. We use multiple times the function, so... probably makes sense to have it. But I would rename it to getUserStatusOrDefault and leave as something like:
```go
func (a *App) getUserStatusOrDefault(id string) {
    status, err := a.GetStatus(id)
    if err != nil {
        return &model.Status{UserId, id, Status: model.STATUS_OFFLINE, manual: false, lastActivity: 0, activeChannel: ""}
    }
    return status
}
```

Reply:
I wasnt if these were the defaults or just the values for these 2-3 particular cases thats why i l loaded all these arguments. Having defaults makes sense now 😄